### PR TITLE
first_attem

### PR DESCRIPTION
**Pull request by neotouch1**
A plugin was made that replaces the addition and multiplication instructions with fma

### CODE REVIEW

File: llvm/lib/Target/X86/lab3/kazantsev_evgeny/X86KazantsevPass.cpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+bool X86KazantsevPass::runOnMachineFunction(MachineFunction &MF) {
+  bool Modified = false;
+  const TargetInstrInfo *TII = MF.getSubtarget().getInstrInfo();
+
+  for (auto &MBB : MF) {
+    std::vector<llvm::MachineBasicBlock::iterator> mulToErase;
+    llvm::Register ar2;
+    MachineInstr *AI;
+    for (auto I = MBB.begin(); I != MBB.end(); I++) {
+      if (I->getOpcode() == X86::MULPDrr) {
```
Comment (m-ly4) [inline llvm/lib/Target/X86/lab3/kazantsev_evgeny/X86KazantsevPass.cpp]:
```suggestion
      if (I->getOpcode() != X86::MULPDrr) {
        continue;
      }
```

File: llvm/lib/Target/X86/lab3/kazantsev_evgeny/X86KazantsevPass.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        const llvm::Register mr0 = I->getOperand(0).getReg();
+        bool Found = false;
+        for (auto J = I; J != MBB.end(); J++) {
+          if (J->getOpcode() == X86::ADDPDrr) {
```
Comment (m-ly4) [inline llvm/lib/Target/X86/lab3/kazantsev_evgeny/X86KazantsevPass.cpp]:
```suggestion
          if (J->getOpcode() != X86::ADDPDrr) {
            continue;
          }
```

File: llvm/test/lab3/kazantsev_evgeny/Mtr.mir [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+  
+    %3:vr128 = COPY $xmm3
+    %1:vr128 = COPY $xmm1
+    %0:vr128 = COPY $xmm0
+    ; CHECK:        %0:vr128 = COPY $xmm0
+    ; CHECK-NOT:    %4:vr128 = nofpexcept MULPDrr %1, %3, implicit $mxcsr
+    %4:vr128 = nofpexcept MULPDrr %1, %3, implicit $mxcsr
+    ; CHECK:        %5:vr128 = nofpexcept ADDPDrr %0, %3, implicit $mxcsr
+    ; CHECK-NEXT:   %6:vr128 = VFMADD213PDr %1, %3, %5, implicit $mxcsr
+    ; CHECK-NEXT:   %7:vr128 = nofpexcept MULPDrr %5, killed %6, implicit $mxcsr
```
Comment (m-ly4) [inline llvm/test/lab3/kazantsev_evgeny/Mtr.mir]: What if instruction will be the following:
```suggestion
    ; CHECK-NEXT:   %7:vr128 = nofpexcept MULPDrr %5, killed %4, implicit $mxcsr
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/lab3/kazantsev_evgeny/Mtr.mir (hunk 3 above)
Comment (neotouch1) [reply llvm/test/lab3/kazantsev_evgeny/Mtr.mir]: The test assumes that the VFMADD213PDr instruction (previous line) will be followed by a MULPDrr instruction and that the register specified as %6 will be destroyed or corrupted in this instruction. However, the kill attribute actually points to register %4, which is not the expected string in the test.

File: llvm/lib/Target/X86/lab3/kazantsev_evgeny/X86KazantsevPass.cpp (hunk 1 above)
Comment (neotouch1) [reply llvm/lib/Target/X86/lab3/kazantsev_evgeny/X86KazantsevPass.cpp]: Yes, I agree here.

File: llvm/lib/Target/X86/lab3/kazantsev_evgeny/X86KazantsevPass.cpp (hunk 2 above)
Comment (neotouch1) [reply llvm/lib/Target/X86/lab3/kazantsev_evgeny/X86KazantsevPass.cpp]: And here

**Review by m-ly4 — APPROVED** [review]

File: llvm/test/lab3/kazantsev_evgeny/Mtr.mir (hunk 3 above)
Comment (m-ly4) [reply llvm/test/lab3/kazantsev_evgeny/Mtr.mir]: Ok, nevermind.

**Pull request merged**